### PR TITLE
ListView CSS updates

### LINKED ISSUE #2266: ListView CSS updates
emphasized / more exact styling as per updated Spectrum guidelines

### PR DESCRIPTION
**Pull request by reidbarber**
Closes #2266

Fixes some existing styles, as well as adds ~~`isEmphasized` support~~ emphasized styles by default.

Changes:
- ~~fixes default "selected" style to have grey background (not blue)~~
- ~~fixes checkboxes to only be blue (emphasized) if ListView is emphasized~~
- ~~add isEmphasized prop and emphasized styles to ListView~~
- add emphasized styles to list items by default
- add emphasized styles to checkboxes
- fix quiet hover border radius styles
- fix border radius for first and last items to match container if selected See comment: https://github.com/adobe/react-spectrum/pull/2517#issuecomment-962097084
- fix 'selectionStyle: highlight' focus ring top border

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Check ListView storybook stories.

## 🧢 Your Project:

RSP

### CODE REVIEW

**Comment by reidbarber** [comment]
One open question:

Currently, the borders of the first and last items in the list look a bit off especially when selected (due to the ListView's border radius):
Screen Shot 2021-11-05 at 12 39 27 PM
Screen Shot 2021-11-05 at 12 35 59 PM

Adding the same border radius to the first and last item looks a bit better, but they won't match so maybe that isn't the best option: 

Screen Shot 2021-11-05 at 12 36 39 PM
Screen Shot 2021-11-05 at 12 36 28 PM

Can anyone suggest a better solution? Not sure if we've dealt with similar problems in other components or not.

**Comment by ktabors** [comment]
> Currently, the borders of the first and last items in the list look a bit off especially when selected (due to the ListView's border radius):

image

I thinking I'm looking at the current build and I don't like how this looks. What do you mean it doesn't match the top?

**Review by ktabors — COMMENTED** [review]
We we have a test to check that isEmphasized class is added?

File: packages/@react-spectrum/list/src/listview.css [hunk 1]
```diff
@@ -87,12 +98,15 @@
...
-    background-color: var(--spectrum-table-quiet-row-background-color-selected-hover);
+  
+  /* give first and last items border-radius to match listview container */
+  div:first-child > div[role="row"] > & {
+    border-start-start-radius: var(--spectrum-listview-item-start-end-border-radius);
+    border-start-end-radius: var(--spectrum-listview-item-start-end-border-radius);
+  }
+  div:last-child > div[role="row"] > & {
+    border-end-start-radius: var(--spectrum-listview-item-start-end-border-radius);
+    border-end-end-radius: var(--spectrum-listview-item-start-end-border-radius);
```
Comment (reidbarber) [inline packages/@react-spectrum/list/src/listview.css:109]: I think this is what we want. Gives the first and last items a border radius that is the ListView's border radius minus the ListView's border width:

Screen Shot 2021-11-05 at 2 18 14 PM

And this selector is a bit awkward so if anyone can recommend a better one, let me know.

File: packages/@react-spectrum/list/src/listview.css [hunk 2]
```diff
@@ -50,31 +53,39 @@
     border-bottom: 0;
   }
 
+  .react-spectrum-ListView--emphasized & {
+    &.is-selected {
+      background-color: var(--spectrum-table-row-background-color-selected);
+      border: 1px solid var(--spectrum-global-color-blue-500);
```
Comment (snowystinger) [inline packages/@react-spectrum/list/src/listview.css:59]: border between items is doubling up, maybe omit the border bottom on all but the last item?

**Review by snowystinger — COMMENTED** [review]
Not something you can do much about, posting here for design really. But if you have scrollbars on, then rounding the top right and bottom right on the first/last elements respectively looks a bit odd
Screen Shot 2021-11-05 at 1 30 54 PM
Screen Shot 2021-11-05 at 1 29 13 PM
Screen Shot 2021-11-05 at 1 28 53 PM

File: packages/@react-spectrum/list/src/listview.css (hunk 2 above)
Comment (reidbarber) [reply packages/@react-spectrum/list/src/listview.css:59]: Pushed a fix to this in https://github.com/adobe/react-spectrum/pull/2517/commits/648711a58ad8d944afdeb6c14e0ed1697109f70f. It is not css-only, but I am open to changing it to be. The selector would be messy with all the divs added by virtualizer.

File: packages/@react-spectrum/list/src/listview.css (hunk 1 above)
Comment (snowystinger) [reply packages/@react-spectrum/list/src/listview.css:113]: selector seems fine, it's traverses right to left, so the list item is already pretty specific, shouldn't cause any performance issues
i didn't realize that the items literally swapped every time one scrolls out of view, so this rounded border is technically always applied to the last element in the view, which is great, even while scrolling

File: packages/@react-spectrum/list/src/listview.css (hunk 2 above)
Comment (snowystinger) [reply packages/@react-spectrum/list/src/listview.css:59]: selecting causes the checkbox to shift over by 1px, we'll need to adjust for that https://reactspectrum.blob.core.windows.net/reactspectrum/38a2ac86506dfab52ccdaf8703f90b72e03f754f/storybook-17/index.html?path=/story/listview--selection-single-checkbox-isemphasized
I think it'd be best to implement the border as an inset box shadow, this way when you focus the row, you don't get a 3px border and you also avoid the content shifting

**Review by majornista — CHANGES_REQUESTED** [review]
With highlight selection, the selected style is does not have sufficient contrast to meet [Success Criterion 1.4.11 Non-text Contrast](https://www.w3.org/TR/WCAG21/#non-text-contrast). Which expects a 3:1 contrast ratio between the selected or not-selected states. The selected and focused, not-selected states have the same background color.  So it is impossible to detect the state change if you press Alt+Space to deselect the focused row.

When we display a checkbox to indicate selection, this is not a problem. But, without some other affordance to distinguish the selected state, we will need a more pronounced color change between the background and foreground. You will notice that for the MacOS finder, the selected row is a more saturated blue with a white background.

File: packages/@react-spectrum/list/src/listview.css (hunk 2 above)
Comment (reidbarber) [reply packages/@react-spectrum/list/src/listview.css]: Pushed a fix via https://github.com/adobe/react-spectrum/pull/2517/commits/a85a767486a6971e2ddcf7eeaf8be8a0116c1b44, let me know what you think.

**Review by devongovett — APPROVED** [review]

**Comment by devongovett** [comment]
Hmm not sure if the checkbox selection is supposed to have the blue border around items also. In TableView it does not. 
image

**Review by dannify — APPROVED** [review]

File: packages/@react-spectrum/list/src/listview.css [hunk 3]
```diff
@@ -33,7 +36,8 @@
 
 .react-spectrum-ListViewItem {
   display: grid; /* TODO: define grid areas */
-  border-bottom: 1px solid var(--spectrum-tabel-cell-border-color, var(--spectrum-global-color-gray-300));
+  border: 1px solid transparent;
+  border-bottom-color: var(--spectrum-tabel-cell-border-color, var(--spectrum-global-color-gray-300));
```
Comment (LFDanLu) [inline packages/@react-spectrum/list/src/listview.css:40]:
```suggestion
  border-bottom-color: var(--spectrum-table-cell-border-color, var(--spectrum-global-color-gray-300));
```
Unless this was on purpose to force the gray fallback?

**Comment by reidbarber** [comment]
> Hmm not sure if the checkbox selection is supposed to have the blue border around items also. In TableView it does not. image

@devongovett I think this is correct, the latest designs have both:

ListView: NOT QUIET (STANDARD) Multiple selection

(from Standard Multiple Selection)

Lmk if this should be changed.

File: packages/@react-spectrum/list/src/listview.css (hunk 3 above)
Comment (reidbarber) [reply packages/@react-spectrum/list/src/listview.css]: Nope, great catch. Definitely just a typo

**Review by LFDanLu — APPROVED** [review]

**Review by dannify — APPROVED** [review]

**Comment by dannify** [comment]
> @devongovett I think this is correct, the latest designs have both:
> 
> ListView: NOT QUIET (STANDARD) Multiple selection
> 
> (from Standard Multiple Selection)
> 
> Lmk if this should be changed.

We can confirm with design. Probably a few versions floating around.

**Pull request merged**